src/plugins/kibana/public/discover/controllers/discover.js

```diff
@@ -362,30 +363,26 @@ define(function (require) {
             return failure.index + failure.shard + failure.reason;
           });
         }
+      }));
 
-        var rows = $scope.rows;
-        var indexPattern = $scope.searchSource.get('index');
+      segmented.on('mergedSegment', function (merged) {
+        $scope.mergedEsResp = merged;
+        $scope.hits = merged.hits.total;
 
-        // merge the rows and the hits, use a new array to help watchers
-        rows = $scope.rows = rows.concat(resp.hits.hits);
+        var indexPattern = $scope.searchSource.get('index');
 
-        if (sortFn) {
-          notify.event('resort rows', function () {
-            rows.sort(sortFn);
-            rows = $scope.rows = rows.slice(0, totalSize);
-            $scope.fieldCounts = {};
-          });
-        }
+        // the merge rows, use a new array to help watchers
+        $scope.rows = merged.hits.hits.slice();
 
         notify.event('flatten hit and count fields', function () {
-          var counts = $scope.fieldCounts;
+          var counts = $scope.fieldCounts = (sortFn ? {} : $scope.fieldCounts) || {};
```

Comment:
Doesn't need to block this PR, but this line is pretty complicated - two assignments, 2 conditionals, 3 branches. In the future we should try to err on the side of multiple lines for something like this.